Performance issue with BuilderXML

The example `DalitzAnalysis_D03K`(200k MC events, 1000 signal events) has a total runtime of 7min from which about 5min are spend on building the intensity. Most of the time is spend in `ComPWA::Physics::updateDataContainerContent`. @spflueger could you think of ways to improve this?

Ok, so when I move the container filling to then end, and only do it once, the FunctionTree prints this error message several times.

> [ERROR] Inverse::execute() | Division by zero

But things are still working. So you get the same result and everything. The runtime decreases from 3:40min to 1:45 min (on my machine).

I don't understand why the FunctionTree evaluates the tree at this point. I don't think its necessary and I would assume that this can be removed. Then the division by zero error message would disappear and I would say this solves the problem.

--- a/Physics/BuilderXML.cpp
+++ b/Physics/BuilderXML.cpp
@@ -59,6 +59,8 @@ IntensityBuilderXML::createIntensity() {
   std::shared_ptr<ComPWA::FunctionTree::FunctionTree> FT =
       createIntensityFT(ModelTree, Data.Data, "");
 
+  updateDataContainerContent();
+
   return {FT, Parameters, Data.Data};
 }
 
@@ -278,7 +280,6 @@ IntensityBuilderXML::createStrengthIntensityFT(
   std::shared_ptr<ComPWA::FunctionTree::FunctionTree> x =
       createIntensityFT(UndecoratedIntensityPT, DataSample, suffix);
 
-  x->parameter();
   tr->insertTree(x, NodeName);
 
   return tr;
@@ -371,7 +372,6 @@ IntensityBuilderXML::createIntegrationStrategyFT(
   if (IntegratorClassName == "MCIntegrationStrategy") {
     // update the PhspData container
     updateDataContainerState();
-    updateDataContainerContent();
 
     std::string NodeName =
         "Normalization(" + UnnormalizedIntensity->Head->name() + ")";
@@ -393,8 +393,6 @@ IntensityBuilderXML::createIntegrationStrategyFT(
     if (PhspWeights)
       tr->createLeaf("EventWeight", PhspWeights, "WeightedIntensities");
     tr->insertTree(UnnormalizedIntensity, "WeightedIntensities");
-
-    tr->parameter();
   } else {
     LOG(WARNING) << "IntensityBuilderXML::createIntegrationStrategyFT(): "
                     "IntegrationStrategy type "
@@ -606,7 +604,6 @@ IntensityBuilderXML::createSequentialAmplitudeFT(
       std::shared_ptr<ComPWA::FunctionTree::FunctionTree> AmpTree =
           createAmplitudeFT(v.second, DataSample, suffix);
 
-      AmpTree->parameter();
       ampname += AmpTree->Head->name();
       Amplitudes.push_back(AmpTree);
     } else if (v.first != "<xmlattr>") {
@@ -835,8 +832,6 @@ IntensityBuilderXML::createHelicityDecayFT(
     tr->insertTree(ProductionFormFactorFT, nodeName);
   }
 
-  tr->parameter();
-
   return tr;
 }
 
@@ -866,6 +861,7 @@ void IntensityBuilderXML::updateDataContainerState() {
 void updateDataContainerContent(ComPWA::FunctionTree::ParameterList &DataList,
                                 const std::vector<ComPWA::Event> &DataSample,
                                 const Kinematics &Kin) {
+  LOG(INFO) << "Updating data container content...";
   auto DataSet = ComPWA::Data::convertEventsToDataSet(DataSample, Kin);
 
   // just loop over the vectors and fill in the data

